### PR TITLE
[Confluence] Add spaces as a configurable field

### PR DESCRIPTION
**Pull request by praveen-elastic**
## Closes https://github.com/elastic/connectors-py/issues/659

This PR contains an enhancement suggested by Dana in a Technical Documentation. It introduces a new configuration parameter named `spaces`.

Following Changes have been added:
- Space in configuration parameters
- Default value of * to index all spaces and its content
- method to verify configured spaces are available
- corresponding tests for the method to verify spaces

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

### CODE REVIEW

File: connectors/sources/confluence.py [hunk 1]
```diff
@@ -37,12 +37,12 @@
 SPACE_QUERY = "limit=100"
 ATTACHMENT_QUERY = "limit=100&expand=version"
 CONTENT_QUERY = (
-    "limit=100&expand=children.attachment,history.lastUpdated,body.storage,space"
+    "limit=50&expand=children.attachment,history.lastUpdated,body.storage,space"
```
Comment (artem-shelkovnikov) [inline connectors/sources/confluence.py:40]: Why change the limit to 50?

File: connectors/sources/confluence.py [hunk 2]
```diff
@@ -37,12 +37,12 @@
...
 )
 
 URLS = {
     SPACE: "rest/api/space?{api_query}",
-    CONTENT: "rest/api/content?{api_query}",
+    CONTENT: "rest/api/content/search?{api_query}",
```
Comment (artem-shelkovnikov) [inline connectors/sources/confluence.py:45]: What about this change too?

Comment (praveen-elastic) [reply connectors/sources/confluence.py:45]: The search endpoint allows us to filter the contents based on spaces.

File: connectors/sources/confluence.py (hunk 1 above)
Comment (praveen-elastic) [reply connectors/sources/confluence.py:40]: We are now using search endpoint and the max allowed results for  search endpoint is 50.

File: connectors/sources/confluence.py [hunk 3]
```diff
@@ -177,6 +178,25 @@ async def paginated_api_call(self, url_name, **url_kwargs):
                 )
                 break
 
+    async def verify_spaces(self):
```
[the remark this answers is not in the document]
Comment (praveen-elastic) [reply connectors/sources/confluence.py:181]: Did you mean I need to override a method from framework for custom validation and move this logic to that method ? I don't see currently any method in the framework.

[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply connectors/sources/confluence.py:181]: Indeed, that's the way - the method is called `validate_config`, you can take it from any other connector as an example.

[the remark this answers is not in the document]
Comment (praveen-elastic) [reply connectors/sources/confluence.py:181]: I know, but the method `validate_config` is executed before ping. This method `verify_spaces` does an API call to the Confluence instance, so logically this verification should happen after ping.

[the remark this answers is not in the document]
Comment (praveen-elastic) [reply connectors/sources/confluence.py:181]: Hey, apologies for the misunderstanding. I agree all 3 ping, validate and get_docs should be independent.
I only wanted to convey that `verify_spaces` method is actually a part of get_docs itself as it is fetching all the spaces which is a sub task of get_docs. The method get_docs fetches spaces, pages/blogs and attachments.

File: connectors/sources/confluence.py [hunk 4]
```diff
@@ -177,6 +178,25 @@ async def paginated_api_call(self, url_name, **url_kwargs):
...
+            return
+        space_keys = []
+        async for response in self.paginated_api_call(
+            url_name=SPACE, api_query=SPACE_QUERY
+        ):
+            spaces = response.get("results", [])
+            space_keys.extend([space["key"] for space in spaces])
+        if unavailable_spaces := set(self.spaces) - set(space_keys):
+            raise Exception(
+                f"Configured unavailable spaces: {', '.join(unavailable_spaces)}"
```
Comment (artem-shelkovnikov) [inline connectors/sources/confluence.py:197]: I would also change the error to something like.

Spaces "{', '.join(unavailable_spaces)}" are not available. Available spaces are: {', '.join(space_keys)}

File: connectors/sources/confluence.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply connectors/sources/confluence.py:181]: Oh no worries. Both `verify_spaces` and `get_docs` can use same methods to fetch spaces, just that `verify_spaces` should provide user with actionable error. `get_docs` though is not supposed to raise errors if user entered wrong configuration.

File: connectors/sources/tests/test_confluence.py [hunk 5]
```diff
@@ -298,6 +315,38 @@ async def test_close_without_client_session():
     assert source.confluence_client.session is None
 
 
+@pytest.mark.asyncio
+async def test_verify_spaces_when_space_keys_are_valid():
+    """Test verify_spaces method with valid space keys"""
```
Comment (timgrein) [inline connectors/sources/tests/test_confluence.py:320]: nit: 
```suggestion
```

IMO we should remove those comments, if they're basically having the same information as the test function name

File: connectors/sources/tests/test_confluence.py [hunk 6]
```diff
@@ -298,6 +315,38 @@ async def test_close_without_client_session():
...
+        return_value=JSONAsyncMock(RESPONSE_SPACE_KEYS)
+    )
+
+    with mock.patch("aiohttp.ClientSession.get", return_value=async_response):
+        await source.confluence_client.verify_spaces()
+
+
+@pytest.mark.asyncio
+async def test_verify_spaces_when_space_keys_are_unavailable_then_raise_exception():
+    """Test verify_spaces method with unavailable space keys"""
```
Comment (timgrein) [inline connectors/sources/tests/test_confluence.py:334]:
```suggestion
```

Same as above

File: connectors/sources/confluence.py [hunk 7]
```diff
@@ -524,11 +554,19 @@ async def get_docs(self, filtering=None):
         Yields:
             dictionary: dictionary containing meta-data of the content.
         """
+        await self.confluence_client.verify_spaces()
```
Comment (timgrein) [inline connectors/sources/confluence.py:557]: As @artem-shelkovnikov already pointed I think we should remove verification logic from `get_docs`. At this point all verification/validations should be done already IMO. You could simply move that logic to `validate_config`, right? Then you can execute the declarative validation first and the "remote validation" afterwards (see [MySQL remote validation](https://github.com/elastic/connectors-python/blob/main/connectors/sources/mysql.py#L393) as an example)

File: connectors/sources/confluence.py [hunk 8]
```diff
@@ -524,11 +554,19 @@ async def get_docs(self, filtering=None):
...
+        if self.confluence_client.spaces == ["*"]:
+            cql = "cql=type="
+        else:
+            quoted_spaces = "','".join(self.confluence_client.spaces)
+            cql = f"cql=space in ('{quoted_spaces}') AND type="
```
Comment (timgrein) [inline connectors/sources/confluence.py:562]: nit: I assume `cql` stands for Confluence Query Language, but what's the actual query here? Is it used to fetch content from configured spaces? I would use a more detailed variable name to make it easier to read.

**Review by timgrein — COMMENTED** [review]
Some nits around tests and variable names and one suggestion around validation.

**Comment by praveen-elastic** [comment]
buildkite test this

File: connectors/sources/confluence.py [hunk 9]
```diff
@@ -304,6 +312,34 @@ def tweak_bulk_options(self, options):
...
+        """Validates whether user input is empty or not for configuration fields.
+        Also validate, if user configured spaces are available in Confluence.
+
+        Raises:
+            Exception: Configured keys can't be empty
+        """
+        self.configuration.check_valid()
+        await self._remote_validation()
+
+    async def _remote_validation(self):
```
Comment (timgrein) [inline connectors/sources/confluence.py:325]: I think we should add `@retryable` to the remove validation to have consistent behaviour across the different connectors.

File: connectors/sources/confluence.py [hunk 10]
```diff
@@ -304,6 +312,34 @@ def tweak_bulk_options(self, options):
...
+            async for response in self.confluence_client.paginated_api_call(
+                url_name=SPACE, api_query=SPACE_QUERY
+            ):
+                spaces = response.get("results", [])
+                space_keys.extend([space["key"] for space in spaces])
+            if unavailable_spaces := set(self.spaces) - set(space_keys):
+                raise ConfigurableFieldValueError(
+                    f"Spaces '{', '.join(unavailable_spaces)}' are not available. Available spaces are: '{', '.join(space_keys)}'"
+                )
+        except Exception:
```
Comment (timgrein) [inline connectors/sources/confluence.py:339]: I don't think we need that `except` block here as every exception from `validate_config` are caught on framework level and logged already (see: [Sync Job Runner](https://github.com/elastic/connectors-python/blob/main/connectors/sync_job_runner.py#L161))

File: connectors/sources/confluence.py (hunk 9 above)
Comment (timgrein) [reply connectors/sources/confluence.py:325]: Makes sense, missed that `api_call` already uses `retryable`.

**Review by timgrein — APPROVED** [review]
LGTM, thank you for addressing the suggestions

**Pull request merged**